File: backend/src/controllers/progressController.js
const User = require('../models/User');

// @desc    Obtenir la progression de l'utilisateur
// @route   GET /api/progress
// @access  Private
exports.getProgress = async (req, res) => {
  try {
    const user = await User.findById(req.user.id);

    res.status(200).json({
      success: true,
      data: {
        completedSteps: user.completedSteps,
        xp: user.xp,
        level: user.level,
        badges: user.badges,
        quizScores: user.quizScores,
        streak: user.streak
      }
    });
  } catch (error) {
    console.error('Erreur getProgress:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur serveur'
    });
  }
};

// @desc    Compléter une étape NIRD
// @route   POST /api/progress/step/:stepId
// @access  Private
exports.completeStep = async (req, res) => {
  try {
    const stepId = parseInt(req.params.stepId);
    const user = await User.findById(req.user.id);

    // Vérifier si l'étape est valide (1-5)
    if (stepId < 1 || stepId > 5) {
      return res.status(400).json({
        success: false,
        message: 'Étape invalide'
      });
    }

    // Vérifier si l'étape n'est pas déjà complétée
    if (user.completedSteps.includes(stepId)) {
      return res.status(400).json({
        success: false,
        message: 'Cette étape est déjà complétée'
      });
    }

    // Ajouter l'étape et l'XP
    user.completedSteps.push(stepId);
    const xpGained = 200;
    await user.addXP(xpGained);

    // Vérifier les badges à débloquer
    const newBadges = [];
    
    if (user.completedSteps.length === 1 && !user.badges.find(b => b.badgeId === 'pioneer')) {
      user.badges.push({ badgeId: 'pioneer' });
      newBadges.push('pioneer');
    }
    
    if (user.completedSteps.length >= 3 && !user.badges.find(b => b.badgeId === 'explorer')) {
      user.badges.push({ badgeId: 'explorer' });
      newBadges.push('explorer');
    }
    
    if (user.completedSteps.length === 5 && !user.badges.find(b => b.badgeId === 'master')) {
      user.badges.push({ badgeId: 'master' });
      newBadges.push('master');
    }

    // Badge spécifique à l'étape sécurité
    if (stepId === 4 && !user.badges.find(b => b.badgeId === 'secure')) {
      user.badges.push({ badgeId: 'secure' });
      newBadges.push('secure');
    }

    // Badge spécifique à l'étape durabilité
    if (stepId === 5 && !user.badges.find(b => b.badgeId === 'eco-warrior')) {
      user.badges.push({ badgeId: 'eco-warrior' });
      newBadges.push('eco-warrior');
    }

    // Mettre à jour le streak
    await user.updateStreak();
    await user.save();

    res.status(200).json({
      success: true,
      message: `Étape ${stepId} complétée! +${xpGained} XP`,
      data: {
        completedSteps: user.completedSteps,
        xp: user.xp,
        level: user.level,
        xpGained,
        newBadges,
        badges: user.badges,
        streak: user.streak
      }
    });
  } catch (error) {
    console.error('Erreur completeStep:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur serveur'
    });
  }
};

// @desc    Annuler une étape NIRD
// @route   DELETE /api/progress/step/:stepId
// @access  Private
exports.uncompleteStep = async (req, res) => {
  try {
    const stepId = parseInt(req.params.stepId);
    const user = await User.findById(req.user.id);

    // Vérifier si l'étape est complétée
    if (!user.completedSteps.includes(stepId)) {
      return res.status(400).json({
        success: false,
        message: 'Cette étape n\'est pas complétée'
      });
    }

    // Retirer l'étape
    user.completedSteps = user.completedSteps.filter(id => id !== stepId);
    user.xp = Math.max(0, user.xp - 200);
    user.calculateLevel();
    await user.save();

    res.status(200).json({
      success: true,
      message: `Étape ${stepId} retirée`,
      data: {
        completedSteps: user.completedSteps,
        xp: user.xp,
        level: user.level
      }
    });
  } catch (error) {
    console.error('Erreur uncompleteStep:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur serveur'
    });
  }
};

// @desc    Sauvegarder un score de quiz
// @route   POST /api/progress/quiz
// @access  Private
exports.saveQuizScore = async (req, res) => {
  try {
    const { stage, score, totalQuestions } = req.body;
    const user = await User.findById(req.user.id);

    // Ajouter le score
    user.quizScores.push({
      stage,
      score,
      totalQuestions
    });

    // Calculer XP bonus basé sur le score
    const percentage = (score / totalQuestions) * 100;
    let xpGained = stage * 10 * score; // XP de base
    
    if (percentage === 100) {
      xpGained += 100; // Bonus parfait
    } else if (percentage >= 80) {
      xpGained += 50; // Bonus excellent
    }

    await user.addXP(xpGained);

    // Badge Quiz Master si tous les stages complétés avec >80%
    const allStagesCompleted = [1, 2, 3, 4, 5].every(s => {
      const stageScores = user.quizScores.filter(qs => qs.stage === s);
      return stageScores.some(qs => (qs.score / qs.totalQuestions) >= 0.8);
    });

    let newBadge = null;
    if (allStagesCompleted && !user.badges.find(b => b.badgeId === 'quiz-master')) {
      user.badges.push({ badgeId: 'quiz-master' });
      newBadge = 'quiz-master';
    }

    await user.updateStreak();
    await user.save();

    res.status(200).json({
      success: true,
      message: `Quiz stage ${stage} sauvegardé! +${xpGained} XP`,
      data: {
        quizScores: user.quizScores,
        xp: user.xp,
        level: user.level,
        xpGained,
        newBadge,
        streak: user.streak
      }
    });
  } catch (error) {
    console.error('Erreur saveQuizScore:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur serveur'
    });
  }
};

// @desc    Obtenir le classement
// @route   GET /api/progress/leaderboard
// @access  Public
exports.getLeaderboard = async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 10;

    const users = await User.find()
      .select('name role xp level badges completedSteps')
      .sort({ xp: -1 })
      .limit(limit);

    res.status(200).json({
      success: true,
      count: users.length,
      data: users.map((user, index) => ({
        rank: index + 1,
        name: user.name,
        role: user.role,
        xp: user.xp,
        level: user.level,
        badgeCount: user.badges.length,
        completedSteps: user.completedSteps.length
      }))
    });
  } catch (error) {
    console.error('Erreur getLeaderboard:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur serveur'
    });
  }
};

// @desc    Obtenir les statistiques globales
// @route   GET /api/progress/stats
// @access  Public
exports.getGlobalStats = async (req, res) => {
  try {
    const totalUsers = await User.countDocuments();
    const totalXP = await User.aggregate([
      { $group: { _id: null, total: { $sum: '$xp' } } }
    ]);
    const avgCompletion = await User.aggregate([
      { $project: { completedCount: { $size: '$completedSteps' } } },
      { $group: { _id: null, avg: { $avg: '$completedCount' } } }
    ]);

    res.status(200).json({
      success: true,
      data: {
        totalUsers,
        totalXP: totalXP[0]?.total || 0,
        avgCompletion: Math.round((avgCompletion[0]?.avg || 0) * 20), // En pourcentage
        totalSteps: 5
      }
    });
  } catch (error) {
    console.error('Erreur getGlobalStats:', error);
    res.status(500).json({
      success: false,
      message: 'Erreur serveur'
    });
  }
};

File: backend-spring/src/main/java/com/nird/api/controller/ProgressController.java
package com.nird.api.controller;

import com.nird.api.dto.ApiResponse;
import com.nird.api.model.Badge;
import com.nird.api.model.QuizScore;
import com.nird.api.model.User;
import com.nird.api.repository.UserRepository;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/progress")
@CrossOrigin(origins = "*")
public class ProgressController {
    @Autowired
    private UserRepository userRepository;

    @GetMapping
    public ResponseEntity<ApiResponse<Map<String, Object>>> getProgress(Authentication authentication) {
        User user = (User) authentication.getPrincipal();
        user = userRepository.findById(user.getId()).orElseThrow();

        Map<String, Object> data = new HashMap<>();
        data.put("completedSteps", user.getCompletedSteps());
        data.put("xp", user.getXp());
        data.put("level", user.getLevel());
        data.put("badges", user.getBadges());
        data.put("quizScores", user.getQuizScores());
        data.put("streak", user.getStreak());

        return ResponseEntity.ok(ApiResponse.success(data));
    }

    @PostMapping("/step/{stepId}")
    public ResponseEntity<ApiResponse<Map<String, Object>>> completeStep(
            @PathVariable Integer stepId, Authentication authentication) {
        User user = (User) authentication.getPrincipal();
        user = userRepository.findById(user.getId()).orElseThrow();

        if (stepId < 1 || stepId > 5) {
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error("Étape invalide"));
        }

        if (user.getCompletedSteps().contains(stepId)) {
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error("Cette étape est déjà complétée"));
        }

        user.getCompletedSteps().add(stepId);
        int xpGained = 200;
        user.addXP(xpGained);

        // Vérifier les badges
        List<String> newBadges = checkBadges(user, stepId);

        updateStreak(user);
        user.setUpdatedAt(LocalDateTime.now());
        user = userRepository.save(user);

        Map<String, Object> data = new HashMap<>();
        data.put("completedSteps", user.getCompletedSteps());
        data.put("xp", user.getXp());
        data.put("level", user.getLevel());
        data.put("xpGained", xpGained);
        data.put("newBadges", newBadges);
        data.put("badges", user.getBadges());
        data.put("streak", user.getStreak());

        return ResponseEntity.ok(ApiResponse.success("Étape " + stepId + " complétée! +" + xpGained + " XP", data));
    }

    @PostMapping("/quiz")
    public ResponseEntity<ApiResponse<Map<String, Object>>> saveQuizScore(
            @Valid @RequestBody Map<String, Integer> request, Authentication authentication) {
        User user = (User) authentication.getPrincipal();
        user = userRepository.findById(user.getId()).orElseThrow();

        Integer stage = request.get("stage");
        Integer score = request.get("score");
        Integer totalQuestions = request.get("totalQuestions");

        QuizScore quizScore = new QuizScore();
        quizScore.setStage(stage);
        quizScore.setScore(score);
        quizScore.setTotalQuestions(totalQuestions);
        quizScore.setUser(user);
        quizScore.setCompletedAt(LocalDateTime.now());
        user.getQuizScores().add(quizScore);

        // Calculer XP
        double percentage = (score.doubleValue() / totalQuestions) * 100;
        int xpGained = stage * 10 * score;
        if (percentage == 100) {
            xpGained += 100;
        } else if (percentage >= 80) {
            xpGained += 50;
        }

        user.addXP(xpGained);
        user.getStats().setTotalQuizzesTaken(user.getStats().getTotalQuizzesTaken() + 1);
        user.getStats().setTotalCorrectAnswers(user.getStats().getTotalCorrectAnswers() + score);

        updateStreak(user);
        user.setUpdatedAt(LocalDateTime.now());
        user = userRepository.save(user);

        Map<String, Object> data = new HashMap<>();
        data.put("quizScores", user.getQuizScores());
        data.put("xp", user.getXp());
        data.put("level", user.getLevel());
        data.put("xpGained", xpGained);
        data.put("streak", user.getStreak());

        return ResponseEntity.ok(ApiResponse.success("Quiz stage " + stage + " sauvegardé! +" + xpGained + " XP", data));
    }

    @GetMapping("/leaderboard")
    public ResponseEntity<ApiResponse<List<Map<String, Object>>>> getLeaderboard(
            @RequestParam(defaultValue = "10") Integer limit) {
        List<User> users = userRepository.findTopByOrderByXpDesc(PageRequest.of(0, limit));

        List<Map<String, Object>> leaderboard = users.stream().map((user, index) -> {
            Map<String, Object> entry = new HashMap<>();
            entry.put("rank", index + 1);
            entry.put("name", user.getName());
            entry.put("role", user.getRole().getValue());
            entry.put("xp", user.getXp());
            entry.put("level", user.getLevel());
            entry.put("badgeCount", user.getBadges().size());
            entry.put("completedSteps", user.getCompletedSteps().size());
            return entry;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(ApiResponse.success(leaderboard));
    }

    @GetMapping("/stats")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getGlobalStats() {
        List<User> allUsers = userRepository.findAll();
        
        long totalUsers = allUsers.size();
        long totalXP = allUsers.stream().mapToLong(User::getXp).sum();
        double avgCompletion = allUsers.stream()
                .mapToDouble(u -> (u.getCompletedSteps().size() / 5.0) * 100)
                .average()
                .orElse(0.0);

        Map<String, Object> data = new HashMap<>();
        data.put("totalUsers", totalUsers);
        data.put("totalXP", totalXP);
        data.put("avgCompletion", Math.round(avgCompletion));
        data.put("totalSteps", 5);

        return ResponseEntity.ok(ApiResponse.success(data));
    }

    private List<String> checkBadges(User user, Integer stepId) {
        List<String> newBadges = new java.util.ArrayList<>();
        
        if (user.getCompletedSteps().size() == 1 && !hasBadge(user, "pioneer")) {
            addBadge(user, "pioneer", "Pionnier", "Première étape complétée!", Badge.BadgeRarity.COMMON);
            newBadges.add("pioneer");
        }
        
        if (user.getCompletedSteps().size() >= 3 && !hasBadge(user, "explorer")) {
            addBadge(user, "explorer", "Explorateur", "3 étapes complétées!", Badge.BadgeRarity.RARE);
            newBadges.add("explorer");
        }
        
        if (user.getCompletedSteps().size() == 5 && !hasBadge(user, "master")) {
            addBadge(user, "master", "Maître", "Toutes les étapes complétées!", Badge.BadgeRarity.EPIC);
            newBadges.add("master");
        }
        
        if (stepId == 4 && !hasBadge(user, "secure")) {
            addBadge(user, "secure", "Sécurisé", "Étape sécurité complétée!", Badge.BadgeRarity.RARE);
            newBadges.add("secure");
        }
        
        if (stepId == 5 && !hasBadge(user, "eco-warrior")) {
            addBadge(user, "eco-warrior", "Éco-Guerrier", "Étape durabilité complétée!", Badge.BadgeRarity.EPIC);
            newBadges.add("eco-warrior");
        }
        
        return newBadges;
    }

    private boolean hasBadge(User user, String badgeId) {
        return user.getBadges().stream().anyMatch(b -> b.getBadgeId().equals(badgeId));
    }

    private void addBadge(User user, String badgeId, String name, String description, Badge.BadgeRarity rarity) {
        Badge badge = new Badge();
        badge.setBadgeId(badgeId);
        badge.setName(name);
        badge.setDescription(description);
        badge.setRarity(rarity);
        badge.setUser(user);
        user.getBadges().add(badge);
    }

    private void updateStreak(User user) {
        LocalDateTime now = LocalDateTime.now();
        LocalDateTime lastActivity = user.getStreak().getLastActivity();
        long diffDays = java.time.Duration.between(lastActivity, now).toDays();

        if (diffDays == 0) {
            return;
        } else if (diffDays == 1) {
            user.getStreak().setCurrent(user.getStreak().getCurrent() + 1);
            if (user.getStreak().getCurrent() > user.getStreak().getBest()) {
                user.getStreak().setBest(user.getStreak().getCurrent());
            }
        } else {
            user.getStreak().setCurrent(1);
        }

        user.getStreak().setLastActivity(now);
    }
}

